html/modules/custom/docstore/docstore.module

```diff
@@ -77,6 +82,133 @@ function docstore_search_api_solr_converted_query_alter(QueryInterface $solarium
+    //
+    // @todo there seems to be some `*s_*` versions of the below fields,
+    // remove them?
+    if (isset($document->sm_files)) {
+      $files = [];
+      // When there is only one item, search api (solr?) flatten that to
+      // a scalar instead of keeping it as an array.
+      if (!is_array($document->sm_files)) {
+        $files[] = implode('||', [
+          0,
+          $document->sm_files,
+          $document->sm_files_file_filemime_,
+          $document->itm_files_file_filesize_,
+          $document->sm_files_file_uri_,
+          $document->sm_files_file_uuid_,
+          $document->dm_files_media_changed_,
+          $document->dm_files_media_created_,
+          $document->sm_files_media_name_,
+          $document->itm_files_media_uid_,
+        ]);
+      }
+      else {
+        foreach ($document->sm_files as $index => $media_uuid) {
+          $files[] = implode('||', [
+            $index,
+            $media_uuid,
+            $document->sm_files_file_filemime_[$index],
+            $document->itm_files_file_filesize_[$index],
+            $document->sm_files_file_uri_[$index],
+            $document->sm_files_file_uuid_[$index],
+            $document->dm_files_media_changed_[$index],
+            $document->dm_files_media_created_[$index],
+            $document->sm_files_media_name_[$index],
+            $document->itm_files_media_uid_[$index],
+          ]);
+        }
+      }
+      $document->addField('sm_concatenated_files_', $files);
+    }
+  }
+}
+
+/**
+ * Implements search_api_solr_search_results_alter().
+ *
+ * This update the file fields with the data in the correct order as Solr
+ * doesn't seem to preserve the order of multi-values fields.
+ *
+ * @todo handle entity reference fields and link fields with the same logic.
+ */
+function docstore_search_api_solr_search_results_alter(ResultSetInterface $result_set, SearchApiQueryInterface $query, SolariumResult $result) {
+  $fields = [
+    'files',
+    'files_file_filemime_',
+    'files_file_filesize_',
+    'files_file_uri_',
+    'files_file_uuid_',
+    'files_media_changed_',
+    'files_media_created_',
+    'files_media_name_',
+    'files_media_uid_',
+  ];
+
+  // Get the iteraror for the result set, we will iterator through it at the
+  // same time as we iterate through the solr results which, normally, have the
+  // same number of items as the result set was already pupulated with the
+  // solr data.
+  $iterator = $result_set->getIterator();
+  foreach ($result as $document) {
+    $item = $iterator->current();
+
+    if (isset($document->sm_concatenated_files_)) {
+      $concatenated_files = $document->sm_concatenated_files_;
+      if (!is_array($concatenated_files)) {
+        $concatenated_files = [$concatenated_files];
+      }
+
+      // The concatenate_files start with their delta, sort then to ensure
+      // they are in the correct order.
+      sort($concatenated_files);
```

Comment:
This fails if there are more than 10 files

Reply:
Yes, thanks for picking that up. I changed it in https://github.com/UN-OCHA/docstore-site/pull/83/commits/6b9da1bb1a67eb4ed2bae922e367bd4a8e6e2580.

Reply:
I'll investigate if we can store the full result somewhere, either in solr or a separate table, I did something similar in https://github.com/UN-OCHA/events-site/blob/develop/html/sites/all/modules/custom/events_event/includes/events_event_search_api_store_data.inc and https://github.com/UN-OCHA/events-site/blob/develop/html/sites/all/modules/custom/events_event/events_event.module#L564

Biggest benefit is that solr only needs to return uuid, no longer fields

Reply:
Yeah, that would be good not to have to do that. Currently, this workaround needs to be adapted to handle entity reference and link fields for which we'll probably still want to index label and title for search.

Elasticsearch can store the entire source data without altering it, I wonder if there is such option in solr.

Reply:
So can solr, https://github.com/UN-OCHA/docstore-site/pull/86